Firestore/Source/Model/FSTMutation.mm

```diff
@@ -384,21 +384,30 @@ - (nullable FSTMaybeDocument *)applyTo:(nullable FSTMaybeDocument *)maybeDoc
 
   for (NSUInteger i = 0; i < serverTransformResults.count; i++) {
     const FieldTransform &fieldTransform = self.fieldTransforms[i];
+    FSTFieldValue *previousValue = nil;
+    if ([baseDocument isMemberOfClass:[FSTDocument class]]) {
+      previousValue = [((FSTDocument *)baseDocument) fieldForPath:fieldTransform.path()];
+    }
+
+    FSTFieldValue *transformResult;
+    // The server just sends null as the transform result for array union / remove operations, so
+    // we have to calculate a result the same as we do for local applications.
+    if (fieldTransform.transformation().type() == TransformOperation::Type::ArrayUnion) {
+      auto elements =
+          static_cast<const ArrayTransform &>(fieldTransform.transformation()).elements();
+      transformResult = [self arrayUnionResultWithElements:elements previousValue:previousValue];
```

Comment:
Instead of passing `elements` here why not pass the `fieldTransform`? Something like

```objc
arrayUnionResultForTransform:(const FieldTransform &)transform
               previousValue:(FSTFieldValue *)previousValue;
```

Reply:
Would pulling something like this out help?

```c++
class ArrayTransform : ... {
  static const std::vector<FSTFieldValue*>& Elements(class TransformOperation& op) {
    FIREBASE_ASSERT(op.type() == Type::ArrayUnion || op.type() == Type::ArrayRemove);
    return static_cast<const ArrayTransform &>(op).elements();
  }
};
```

Reply:
Done. Thanks for writing the code for me. :-)  Note, I changed `Elements(class TransformOperation& op)` to `Elements(const TransformOperation& op)` as I assume that was a typo, but if I'm confused, let me know!

Reply:
You're writing with the benefit of a compiler :-).

Please also use in `localTransformResultsWithBaseDocument`

Reply:
Gah, sorry. ☹️  Fixed.